File: 2435-paths-in-matrix-whose-sum-is-divisible-by-k/2435-paths-in-matrix-whose-sum-is-divisible-by-k.go

const mod int = 1_000_000_007

func numberOfPaths(grid [][]int, k int) int {
	m := len(grid)
	n := len(grid[0])

	dp := make([][][]int, m)
	for i := 0; i < m; i++ {
		dp[i] = make([][]int, n)
		for j := 0; j < n; j++ {
			dp[i][j] = make([]int, k)
		}
	}

	dp[0][0][grid[0][0]%k] = 1
	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			for prev := 0; prev < k; prev++ {
				curr := (prev + grid[i][j]) % k
				if i-1 >= 0 {
					dp[i][j][curr] = (dp[i][j][curr] + dp[i-1][j][prev]) % mod
				}
				if j-1 >= 0 {
					dp[i][j][curr] = (dp[i][j][curr] + dp[i][j-1][prev]) % mod
				}
			}
		}
	}

	return dp[m-1][n-1][0]
}

File: 2435-paths-in-matrix-whose-sum-is-divisible-by-k/2435-paths-in-matrix-whose-sum-is-divisible-by-k.cpp
#pragma GCC optimize("O3", "unroll-loops");
static const int __ = []() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    return 0;
}();
using ll = long long;
const int MOD = 1e9 + 7;

class Solution {
   public:
    int numberOfPaths(vector<vector<int>>& grid, int k) {
        int m = grid.size();
        int n = grid[0].size();
        vector<vector<vector<int>>> dp(
            m, vector<vector<int>>(n, vector<int>(k, 0)));

        dp[0][0][grid[0][0] % k] = 1;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                for (int prev = 0; prev < k; prev++) {
                    int curr = (prev + grid[i][j]) % k;
                    if (i - 1 >= 0) {
                        dp[i][j][curr] =
                            ((ll)dp[i][j][curr] + dp[i - 1][j][prev]) % MOD;
                    }
                    if (j - 1 >= 0) {
                        dp[i][j][curr] =
                            ((ll)dp[i][j][curr] + dp[i][j - 1][prev]) % MOD;
                    }
                }
            }
        }

        return dp[m - 1][n - 1][0];
    }
};